modules/importer/src/server/mod.rs

```diff
@@ -91,58 +90,26 @@ impl Server {
 
             // Remove jobs that are finished
             runs.retain(|_, job| !job.is_finished());
+            let count = runs.len();
 
             // Asynchronously fire off new jobs subject to max concurrency
-            let todo: Vec<_> = service
-                .list()
-                .await?
-                .into_iter()
-                .filter(|i| {
-                    !(runs.contains_key(&i.name) || i.data.configuration.disabled || can_wait(i))
-                })
-                .take(self.concurrency - runs.len())
-                .map(|i| {
-                    (
-                        i.name.clone(),
-                        spawn_local(import(runner.clone(), i, service.clone())),
-                    )
-                })
-                .collect();
-
-            // Add them to our list of currently running jobs
-            runs.extend(todo.into_iter());
-        }
-    }
-
-    /// Reset all jobs back into non-running state.
-    ///
-    /// This is intended when the application starts up, to reset stale job states. Making it
-    /// possible to re-run them when they are due.
-    ///
-    /// **NOTE:** we can only do this as we're intended to be a single-process worker.
-    async fn reset_all_jobs(&self, service: &ImporterService) -> anyhow::Result<()> {
-        for importer in service.list().await? {
-            if importer.data.state == State::Running {
-                log::info!(
-                    "Cleaning up stale importer job during startup: {} (since: {})",
-                    importer.name,
-                    importer.data.last_change
-                );
+            runs.extend(
                 service
-                    .update_finish(
-                        &importer.name,
-                        None,
-                        // either use the last run, or fall back to the last time the state changed
-                        importer.data.last_run.unwrap_or(importer.data.last_change),
-                        Some("Import cancelled".into()),
-                        None,
-                        None,
-                    )
-                    .await?;
-            }
+                    .list()
+                    .await?
+                    .into_iter()
+                    .filter(|i| {
+                        !(i.data.configuration.disabled || already_running(i) || can_wait(i))
```

Comment:
I think there's a race condition here. If multiple instances retrieve that at the same time, they would also start those jobs, at the same time. Leading to multiple instances of the same job running.

Reply:
There is absolutely a race condition here. It's accounted for in the very first line of the `import` function. It's entirely possible that a task spawned from this loop may end immediately, before it processes any data.